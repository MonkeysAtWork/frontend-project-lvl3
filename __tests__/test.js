import nock from 'nock';
import path from 'path';
import fs from 'fs';
import { html } from 'js-beautify';

import run from '../src/application';


nock.disableNetConnect();

const fixturesPath = path.join(__dirname, '__fixtures__');


const response = fs.readFileSync(path.join(fixturesPath, 'rss_0.xml'), 'utf-8');
const url = 'http://www.test.com';
const inputEvent = new Event('input', { bubbles: true });
const submitEvent = new Event('submit');
let form;
let input;


const delay = (time) => new Promise((resolve) => setTimeout(resolve, time));
const getTree = () => html(document.body.innerHTML);

const inputValue = (value) => {
  input.value = value;
  input.dispatchEvent(inputEvent);
  return delay(50);
};

const submitForm = () => {
  form.dispatchEvent(submitEvent);
  return delay(300);
};

nock('https://cors-anywhere.herokuapp.com')
  // .log(console.log)
  .get(/undefined$/)
  .replyWithError('some error')
  .get(/404$/)
  .reply(404)
  .get(/nofeed$/)
  .reply(200, '<!doctype html><html><head></head><body></body></html>')
  .get(/feed$/)
  .twice()
  .reply(200, response);


beforeEach(() => {
  const initHtml = fs.readFileSync(path.join(fixturesPath, 'index.html'), 'utf-8');
  document.documentElement.innerHTML = initHtml;
  run();
  form = document.querySelector('#rssAddForm');
  input = form.elements.link;
});


test('invalid URL', () => inputValue('uncorrect')
  .then(() => expect(getTree()).toMatchSnapshot()));


test('uncorrect URL', () => inputValue(`${url}/undefined`)
  .then(() => expect(getTree()).toMatchSnapshot())
  .then(submitForm)
  .then(() => expect(getTree()).toMatchSnapshot()));


test('network error', () => inputValue(`${url}/404`)
  .then(() => expect(getTree()).toMatchSnapshot())
  .then(submitForm)
  .then(() => expect(getTree()).toMatchSnapshot()));


test('no feed on URL', () => inputValue(`${url}/nofeed`)
  .then(submitForm)
  .then(() => expect(getTree()).toMatchSnapshot()));


test('correct URL twice', () => inputValue(`${url}/feed`)
  .then(submitForm)
  .then(() => expect(getTree()).toMatchSnapshot())
  .then(() => inputValue(`${url}/feed`))
  .then(() => expect(getTree()).toMatchSnapshot()));
