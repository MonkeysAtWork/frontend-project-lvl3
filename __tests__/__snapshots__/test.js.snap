// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`correct URL twice 1`] = `
"<div class=\\"jumbotron\\" id=\\"header\\">
    <h1 class=\\"display-5\\">RSS aggregator</h1>
    <div id=\\"formConteiner\\">
        <form class=\\"form-inline\\" id=\\"rssAddForm\\">
            <input type=\\"text\\" name=\\"link\\" class=\\"form-control\\" spellcheck=\\"false\\" placeholder=\\"RSS link\\">
            <button type=\\"submit\\" class=\\"btn\\" disabled=\\"true\\">Add</button>
            <div class=\\"invalid-feedback\\" id=\\"feedback\\"></div>
        </form>
    </div>
</div>
<div class=\\"container-fluid\\">
    <div class=\\"row\\">
        <div class=\\"col\\" id=\\"feeds\\">
            <div class=\\"list-group list-group-flush border-right\\">
                <div class=\\"list-group-item\\">
                    <h5 class=\\"mb-1\\">Lorem ipsum feed for an interval of 1 minutes</h5>
                    <p class=\\"mb-1\\">This is a constantly updating lorem ipsum feed</p>
                </div>
            </div>
        </div>
        <div class=\\"col\\" id=\\"posts\\">
            <ul class=\\"list-group list-group-flush overflow-auto\\">
                <li class=\\"list-group-item\\"><a href=\\"http://example.com/test/1584637680\\">Veniam magna occaecat qui irure mollit ea consequat ex pariatur.</a></li>
                <li class=\\"list-group-item\\"><a href=\\"http://example.com/test/1584637620\\">Do reprehenderit nostrud laborum deserunt aliqua aliqua fugiat magna labore ullamco.</a></li>
                <li class=\\"list-group-item\\"><a href=\\"http://example.com/test/1584637560\\">Qui amet et in pariatur et nostrud quis do id.</a></li>
                <li class=\\"list-group-item\\"><a href=\\"http://example.com/test/1584637500\\">Tempor magna non duis sint deserunt pariatur adipisicing elit.</a></li>
                <li class=\\"list-group-item\\"><a href=\\"http://example.com/test/1584637440\\">Incididunt consequat reprehenderit sit adipisicing voluptate sint consequat.</a></li>
                <li class=\\"list-group-item\\"><a href=\\"http://example.com/test/1584637380\\">Labore aliqua incididunt duis voluptate reprehenderit officia est mollit voluptate velit consectetur nulla velit.</a></li>
                <li class=\\"list-group-item\\"><a href=\\"http://example.com/test/1584637320\\">Voluptate deserunt elit consectetur tempor dolore ut dolor ex ullamco sunt aliquip esse sit.</a></li>
                <li class=\\"list-group-item\\"><a href=\\"http://example.com/test/1584637260\\">Eu adipisicing nostrud tempor voluptate commodo laboris eiusmod voluptate ex.</a></li>
                <li class=\\"list-group-item\\"><a href=\\"http://example.com/test/1584637200\\">Sint anim reprehenderit elit commodo.</a></li>
                <li class=\\"list-group-item\\"><a href=\\"http://example.com/test/1584637140\\">Ea officia veniam dolore adipisicing nulla esse aliqua.</a></li>
            </ul>
        </div>
    </div>
</div>"
`;

exports[`correct URL twice 2`] = `
"<div class=\\"jumbotron\\" id=\\"header\\">
    <h1 class=\\"display-5\\">RSS aggregator</h1>
    <div id=\\"formConteiner\\">
        <form class=\\"form-inline\\" id=\\"rssAddForm\\">
            <input type=\\"text\\" name=\\"link\\" class=\\"form-control is-invalid\\" spellcheck=\\"false\\" placeholder=\\"RSS link\\">
            <button type=\\"submit\\" class=\\"btn\\" disabled=\\"true\\">Add</button>
            <div class=\\"invalid-feedback\\" id=\\"feedback\\">this feed already was added</div>
        </form>
    </div>
</div>
<div class=\\"container-fluid\\">
    <div class=\\"row\\">
        <div class=\\"col\\" id=\\"feeds\\">
            <div class=\\"list-group list-group-flush border-right\\">
                <div class=\\"list-group-item\\">
                    <h5 class=\\"mb-1\\">Lorem ipsum feed for an interval of 1 minutes</h5>
                    <p class=\\"mb-1\\">This is a constantly updating lorem ipsum feed</p>
                </div>
            </div>
        </div>
        <div class=\\"col\\" id=\\"posts\\">
            <ul class=\\"list-group list-group-flush overflow-auto\\">
                <li class=\\"list-group-item\\"><a href=\\"http://example.com/test/1584637680\\">Veniam magna occaecat qui irure mollit ea consequat ex pariatur.</a></li>
                <li class=\\"list-group-item\\"><a href=\\"http://example.com/test/1584637620\\">Do reprehenderit nostrud laborum deserunt aliqua aliqua fugiat magna labore ullamco.</a></li>
                <li class=\\"list-group-item\\"><a href=\\"http://example.com/test/1584637560\\">Qui amet et in pariatur et nostrud quis do id.</a></li>
                <li class=\\"list-group-item\\"><a href=\\"http://example.com/test/1584637500\\">Tempor magna non duis sint deserunt pariatur adipisicing elit.</a></li>
                <li class=\\"list-group-item\\"><a href=\\"http://example.com/test/1584637440\\">Incididunt consequat reprehenderit sit adipisicing voluptate sint consequat.</a></li>
                <li class=\\"list-group-item\\"><a href=\\"http://example.com/test/1584637380\\">Labore aliqua incididunt duis voluptate reprehenderit officia est mollit voluptate velit consectetur nulla velit.</a></li>
                <li class=\\"list-group-item\\"><a href=\\"http://example.com/test/1584637320\\">Voluptate deserunt elit consectetur tempor dolore ut dolor ex ullamco sunt aliquip esse sit.</a></li>
                <li class=\\"list-group-item\\"><a href=\\"http://example.com/test/1584637260\\">Eu adipisicing nostrud tempor voluptate commodo laboris eiusmod voluptate ex.</a></li>
                <li class=\\"list-group-item\\"><a href=\\"http://example.com/test/1584637200\\">Sint anim reprehenderit elit commodo.</a></li>
                <li class=\\"list-group-item\\"><a href=\\"http://example.com/test/1584637140\\">Ea officia veniam dolore adipisicing nulla esse aliqua.</a></li>
            </ul>
        </div>
    </div>
</div>"
`;

exports[`invalid URL 1`] = `
"<div class=\\"jumbotron\\" id=\\"header\\">
    <h1 class=\\"display-5\\">RSS aggregator</h1>
    <div id=\\"formConteiner\\">
        <form class=\\"form-inline\\" id=\\"rssAddForm\\">
            <input type=\\"text\\" name=\\"link\\" class=\\"form-control is-invalid\\" spellcheck=\\"false\\" placeholder=\\"RSS link\\">
            <button type=\\"submit\\" class=\\"btn\\" disabled=\\"true\\">Add</button>
            <div class=\\"invalid-feedback\\" id=\\"feedback\\">URL is not correct</div>
        </form>
    </div>
</div>
<div class=\\"container-fluid\\">
    <div class=\\"row\\">
        <div class=\\"col\\" id=\\"feeds\\"></div>
        <div class=\\"col\\" id=\\"posts\\"></div>
    </div>
</div>"
`;

exports[`network error 1`] = `
"<div class=\\"jumbotron\\" id=\\"header\\">
    <h1 class=\\"display-5\\">RSS aggregator</h1>
    <div id=\\"formConteiner\\">
        <form class=\\"form-inline\\" id=\\"rssAddForm\\">
            <input type=\\"text\\" name=\\"link\\" class=\\"form-control\\" spellcheck=\\"false\\" placeholder=\\"RSS link\\">
            <button type=\\"submit\\" class=\\"btn\\">Add</button>
            <div class=\\"invalid-feedback\\" id=\\"feedback\\"></div>
        </form>
    </div>
</div>
<div class=\\"container-fluid\\">
    <div class=\\"row\\">
        <div class=\\"col\\" id=\\"feeds\\"></div>
        <div class=\\"col\\" id=\\"posts\\"></div>
    </div>
</div>"
`;

exports[`network error 2`] = `
"<div class=\\"jumbotron\\" id=\\"header\\">
    <h1 class=\\"display-5\\">RSS aggregator</h1>
    <div id=\\"formConteiner\\">
        <form class=\\"form-inline\\" id=\\"rssAddForm\\">
            <input type=\\"text\\" name=\\"link\\" class=\\"form-control is-invalid\\" spellcheck=\\"false\\" placeholder=\\"RSS link\\">
            <button type=\\"submit\\" class=\\"btn\\" disabled=\\"true\\">Add</button>
            <div class=\\"invalid-feedback\\" id=\\"feedback\\">URL is unreachable</div>
        </form>
    </div>
</div>
<div class=\\"container-fluid\\">
    <div class=\\"row\\">
        <div class=\\"col\\" id=\\"feeds\\"></div>
        <div class=\\"col\\" id=\\"posts\\"></div>
    </div>
</div>"
`;

exports[`no feed on URL 1`] = `
"<div class=\\"jumbotron\\" id=\\"header\\">
    <h1 class=\\"display-5\\">RSS aggregator</h1>
    <div id=\\"formConteiner\\">
        <form class=\\"form-inline\\" id=\\"rssAddForm\\">
            <input type=\\"text\\" name=\\"link\\" class=\\"form-control is-invalid\\" spellcheck=\\"false\\" placeholder=\\"RSS link\\">
            <button type=\\"submit\\" class=\\"btn\\" disabled=\\"true\\">Add</button>
            <div class=\\"invalid-feedback\\" id=\\"feedback\\">No feeds were found for this URL</div>
        </form>
    </div>
</div>
<div class=\\"container-fluid\\">
    <div class=\\"row\\">
        <div class=\\"col\\" id=\\"feeds\\"></div>
        <div class=\\"col\\" id=\\"posts\\"></div>
    </div>
</div>"
`;

exports[`uncorrect URL 1`] = `
"<div class=\\"jumbotron\\" id=\\"header\\">
    <h1 class=\\"display-5\\">RSS aggregator</h1>
    <div id=\\"formConteiner\\">
        <form class=\\"form-inline\\" id=\\"rssAddForm\\">
            <input type=\\"text\\" name=\\"link\\" class=\\"form-control\\" spellcheck=\\"false\\" placeholder=\\"RSS link\\">
            <button type=\\"submit\\" class=\\"btn\\">Add</button>
            <div class=\\"invalid-feedback\\" id=\\"feedback\\"></div>
        </form>
    </div>
</div>
<div class=\\"container-fluid\\">
    <div class=\\"row\\">
        <div class=\\"col\\" id=\\"feeds\\"></div>
        <div class=\\"col\\" id=\\"posts\\"></div>
    </div>
</div>"
`;

exports[`uncorrect URL 2`] = `
"<div class=\\"jumbotron\\" id=\\"header\\">
    <h1 class=\\"display-5\\">RSS aggregator</h1>
    <div id=\\"formConteiner\\">
        <form class=\\"form-inline\\" id=\\"rssAddForm\\">
            <input type=\\"text\\" name=\\"link\\" class=\\"form-control is-invalid\\" spellcheck=\\"false\\" placeholder=\\"RSS link\\">
            <button type=\\"submit\\" class=\\"btn\\" disabled=\\"true\\">Add</button>
            <div class=\\"invalid-feedback\\" id=\\"feedback\\">URL is unreachable</div>
        </form>
    </div>
</div>
<div class=\\"container-fluid\\">
    <div class=\\"row\\">
        <div class=\\"col\\" id=\\"feeds\\"></div>
        <div class=\\"col\\" id=\\"posts\\"></div>
    </div>
</div>"
`;
